File: pluto/src/telem/test/mock.ts
// Copyright 2024 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { observe } from "@synnaxlabs/x";
import { type Mock, vi } from "vitest";

import { type telem } from "@/telem/aether";

class Base {
  cleanupStub: Mock = vi.fn();

  async cleanup(): Promise<void> {
    this.cleanupStub();
  }
}

export class BooleanSink extends Base implements telem.BooleanSink {
  value: boolean | null = null;
  fn: Mock = vi.fn();

  async set(value: boolean): Promise<void> {
    this.value = value;
    this.fn(value);
  }
}

export const booleanSinkSpec: telem.BooleanSinkSpec = {
  type: "boolean-sink",
  props: {},
  variant: "sink",
  valueType: "boolean",
};

export class BooleanSource
  extends observe.Observer<void>
  implements telem.BooleanSource
{
  private iValue: boolean;
  cleanupStub: Mock = vi.fn();

  constructor(value: boolean = false) {
    super();
    this.iValue = value;
  }

  setValue(value: boolean): void {
    this.iValue = value;
    this.notify();
  }

  async value(): Promise<boolean> {
    return this.iValue;
  }

  async cleanup(): Promise<void> {
    this.cleanupStub();
  }
}

export const booleanSourceSpec: telem.BooleanSourceSpec = {
  type: "boolean-source",
  props: {},
  variant: "source",
  valueType: "boolean",
};

const REGISTRY: Record<string, new () => telem.Telem> = {
  "boolean-sink": BooleanSink,
  "boolean-source": BooleanSource,
};

export const FACTORY: telem.Factory = {
  type: "mock",
  create: (spec: telem.Spec): telem.Telem | null => {
    const Constructor = REGISTRY[spec.type];
    return new Constructor();
  },
};

export class Provider implements telem.Provider {
  clusterKey: string = "test";
  key: string = "test";

  equals(): boolean {
    return true;
  }

  created: Record<string, telem.Telem> = {};

  registerFactory(): void {}

  create<T>(spec: telem.Spec): T {
    const t = FACTORY.create(spec) as T;
    this.created[spec.type] = t as unknown as telem.Telem;
    return t;
  }
}

File: core/pkg/service/channel/mock/mock.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

// Package mock provides service-layer channel test helpers for use with a
// distribution-layer mock cluster. It lives in the service layer because it opens a
// real service-layer channel.Service (which the distribution layer must not depend on).
package mock

import (
	"context"
	"sync"

	"github.com/synnaxlabs/synnax/pkg/distribution"
	distmock "github.com/synnaxlabs/synnax/pkg/distribution/mock"
	"github.com/synnaxlabs/synnax/pkg/distribution/framer/writer"
	"github.com/synnaxlabs/synnax/pkg/service/channel"
)

var (
	mu       sync.Mutex
	services = make(map[*distribution.Layer]*channel.Service)
)

// ChannelService returns the service-layer channel.Service for the given mock node,
// opening it (via channel.Wrap, which creates the node's control channel) on first use
// and caching it per node thereafter. Any WrapOptions are applied only on the first call
// for a given node; later calls return the cached service and ignore the options.
func ChannelService(n distmock.Node, opts ...channel.WrapOption) *channel.Service {
	mu.Lock()
	defer mu.Unlock()
	if s, ok := services[n.Layer]; ok {
		return s
	}
	s := channel.Wrap(n.Layer, opts...)
	services[n.Layer] = s
	return s
}

// OpenWriter resolves the distribution-layer channel metadata for cfg.Keys via svc (the
// test's service-layer channel service) and opens a distribution writer against n. Since
// the distribution writer no longer resolves channels itself, service-layer tests use
// this to supply the resolved channels their service-created keys require.
func OpenWriter(
	ctx context.Context,
	n distmock.Node,
	svc *channel.Service,
	cfg writer.Config,
) (*writer.Writer, error) {
	channels, err := svc.RetrieveByKeys(ctx, cfg.Keys...)
	if err != nil {
		return nil, err
	}
	cfg.Channels = channels
	return n.Framer.OpenWriter(ctx, cfg)
}
